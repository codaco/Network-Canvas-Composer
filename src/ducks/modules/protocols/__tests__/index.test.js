/* eslint-env jest */

import { createStore, applyMiddleware } from 'redux';
import { advanceTo } from 'jest-date-mock';
import thunk from 'redux-thunk';
import { isMatch } from 'lodash';
import history from '@app/history';
import reducer, { actionCreators } from '../index';
import { actionCreators as registerActionCreators } from '../register';
import testState from '../../../../__tests__/testState.json';
import { loadProtocolConfiguration } from '../../../../other/protocols';
import { APP_SCHEMA_VERSION } from '../../../../config';

jest.mock('../../../../other/protocols');
jest.mock('../../dialogs');
jest.mock('../../../../config');
jest.mock(
  '../../../../protocol-validation/migrations/migrations',
  () => ([
    { version: 1, migration: protocol => protocol },
    { version: 2, migration: protocol => protocol },
    { version: 3, migration: protocol => protocol },
    { version: 4, migration: protocol => protocol },
  ]),
  { virtual: true },
);

expect.extend({
  containsDialogAction(received, dialogMatcher) {
    const pass = received.some(
      ([action]) =>
        action.type === 'PROTOCOL/OPEN_DIALOG' && isMatch(action.dialog, dialogMatcher),
    );

    if (!pass) {
      return {
        message: () =>
          `expected ${JSON.stringify(received)} to contain dialog action matching ${JSON.stringify(dialogMatcher)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${JSON.stringify(received)} not to contain dialog action matching ${JSON.stringify(dialogMatcher)}`,
      pass: true,
    };
  },
});

expect.extend({
  containsAction(received, actionMatcher) {
    const pass = received.some(action => isMatch(action[0], actionMatcher));

    if (!pass) {
      return {
        message: () =>
          `expected ${JSON.stringify(received)} to contain action matching ${JSON.stringify(actionMatcher)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${JSON.stringify(received)} not to contain action matching ${JSON.stringify(actionMatcher)}`,
      pass: true,
    };
  },
});

const log = jest.fn();

const logger = () =>
  next =>
    (action) => {
      log(action);
      return next(action);
    };

const getStore = (initialState = testState) =>
  createStore(
    (state = initialState) => state,
    applyMiddleware(thunk, logger),
  );

const getProtocol = (mergeProps = {}) => ({
  ...testState.protocol.present,
  ...mergeProps,
});

describe('protocols', () => {
  let store;

  beforeEach(() => {
    log.mockClear();
    store = getStore();
  });

  describe('createAndLoadProtocol()', () => {
    it('triggers create and import actions', () =>
      store.dispatch(actionCreators.createAndLoadProtocol())
        .then(() => {
          expect(log.mock.calls[0]).toEqual([{ payload: 'PROTOCOLS', type: 'STATUS/BUSY' }]);
          expect(log.mock.calls[1]).toEqual([{ type: 'PROTOCOLS/CREATE_PROTOCOL' }]);
          expect(log.mock.calls[2]).toEqual([{
            filePath: '/dev/null/fake/user/entered/path',
            workingPath: '/dev/null/fake/working/path',
            type: 'PROTOCOLS/CREATE_PROTOCOL_SUCCESS',
          }]);
          expect(log.mock.calls[3]).toEqual([{
            advanced: true,
            filePath: '/dev/null/fake/user/entered/path',
            id: '809895df-bbd7-4c76-ac58-e6ada2625f9b',
            type: 'PROTOCOLS/REGISTER',
            workingPath: '/dev/null/fake/working/path',
          }]);

          expect(history.entries.pop()).toMatchObject({
            pathname: '/edit/809895df-bbd7-4c76-ac58-e6ada2625f9b/',
          });
        }),
    );
  });

  describe('saveAndBundleProtocol()', () => {
    advanceTo(Date.UTC(2017, 5, 27, 0, 0, 0));

    describe('invalid protocol', () => {
      beforeEach(() => {
        log.mockClear();
        store = getStore({
          ...testState,
          protocol: {
            present: getProtocol({ stages: [] }),
          },
        });
      });

      it('dispatches an error when protocol is invalid', () =>
        store.dispatch(actionCreators.saveAndBundleProtocol())
          .then(() => {
            expect(log.mock.calls).toMatchSnapshot();
          }),
      );
    });

    it('triggers save and export actions', () =>
      store.dispatch(actionCreators.saveAndBundleProtocol())
        .then(() => {
          expect(log.mock.calls).toMatchSnapshot();
        }),
    );
  });

  describe('unbundleAndLoadProtocol()', () => {
    beforeEach(() => {
      log.mockClear();
    });

    it('triggers import action and load redirect', done =>
      store.dispatch(actionCreators.unbundleAndLoadProtocol('/dev/null/mock/path'))
        .then(() => {
          expect(log.mock.calls).containsAction({ payload: 'PROTOCOLS', type: 'STATUS/BUSY' });
          expect(log.mock.calls).containsAction({ type: 'PROTOCOLS/UNBUNDLE_SUCCESS' });
          expect(log.mock.calls).containsAction({ type: 'PROTOCOLS/REGISTER' });
          expect(history.entries.pop()).toMatchObject({
            pathname: '/edit/809895df-bbd7-4c76-ac58-e6ada2625f9b/',
          });
          done();
        }),
    );

    it('when the protocol is invalid it dispatches an error when protocol is invalid (but still imports it)', () => {
      loadProtocolConfiguration.mockReturnValueOnce(
        Promise.resolve(getProtocol({ schemaVersion: APP_SCHEMA_VERSION, stages: [] })),
      );

      return store.dispatch(actionCreators.unbundleAndLoadProtocol('/dev/null/mock/path/invalid'))
        .then(() => {
          expect(log.mock.calls).containsDialogAction({ type: 'Error' });
          expect(log.mock.calls).toMatchSnapshot();
        });
    });

    it('when the schema version is greater than the app version it notifies user to upgrade the app', () => {
      const futureVersion = APP_SCHEMA_VERSION + 1;

      loadProtocolConfiguration.mockReturnValueOnce(
        Promise.resolve(getProtocol({ schemaVersion: futureVersion })),
      );

      return store.dispatch(actionCreators.unbundleAndLoadProtocol('/dev/null/mock/path/newer-protocol'))
        .then(() => {
          expect(log.mock.calls).toMatchSnapshot();
          expect(log.mock.calls).containsDialogAction({
            title: 'Protocol not compatible with current version',
            type: 'UserError',
          });
        });
    });

    it('when the schema version is less than the app version it attempts to upgrade protocol and then imports it', () => {
      const pastVersion = APP_SCHEMA_VERSION - 1;

      loadProtocolConfiguration.mockReturnValueOnce(
        Promise.resolve(getProtocol({ schemaVersion: pastVersion })),
      );

      return store.dispatch(actionCreators.unbundleAndLoadProtocol('/dev/null/mock/path/older-protocol'))
        .then(() => {
          expect(log.mock.calls).toMatchSnapshot();
          expect(log.mock.calls).containsDialogAction({ title: 'Would you like to upgrade the protocol?', type: 'Confirm' });
        });
    });

    it('when the schema version is the same as the app version it imports the protocol', () => {
      loadProtocolConfiguration.mockReturnValueOnce(
        Promise.resolve(getProtocol()),
      );

      return store.dispatch(actionCreators.unbundleAndLoadProtocol('/dev/null/mock/path/matching-protocol'))
        .then(() => {
          expect(log.mock.calls).toMatchSnapshot();
          expect(log.mock.calls).containsAction({ type: 'PROTOCOLS/UNBUNDLE_SUCCESS' });
        });
    });
  });

  describe('reducer', () => {
    let protocolStore;

    beforeEach(() => {
      protocolStore = createStore(reducer);
    });

    it('REGISTER_PROTOCOL', (done) => {
      protocolStore.subscribe(() => {
        const state = protocolStore.getState();

        expect(state[0]).toMatchObject(
          {
            filePath: '/dev/null/mock/file/path',
            advanced: true,
            id: '809895df-bbd7-4c76-ac58-e6ada2625f9b',
            workingPath: '/dev/null/mock/working/path',
          },
        );

        done();
      });

      protocolStore.dispatch(registerActionCreators.registerProtocol({
        filePath: '/dev/null/mock/file/path',
        workingPath: '/dev/null/mock/working/path',
      }));
    });
  });
});
