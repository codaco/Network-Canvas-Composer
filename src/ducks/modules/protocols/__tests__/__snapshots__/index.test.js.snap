// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protocols saveAndBundleProtocol() invalid protocol dispatches an error when protocol is invalid 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/SAVING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/PREFLIGHT",
    },
  ],
  Array [
    Object {
      "meta": null,
      "protocol": Object {
        "assetManifest": Object {
          "1234-asset-1": Object {
            "source": "1234-4567-asset-1",
          },
          "1234-asset-2": Object {
            "source": "1234-4567-asset-2",
          },
          "1234-asset-3": Object {
            "source": "1234-4567-asset-3",
          },
          "1234-asset-4": Object {
            "source": "1234-4567-asset-4",
          },
          "1234-asset-5": Object {
            "source": "1234-4567-asset-5",
          },
        },
        "codebook": Object {
          "edge": Object {
            "1234-5": Object {
              "color": "blue",
              "name": "an edge",
            },
          },
          "ego": Object {
            "variables": Object {
              "sierra": Object {
                "name": "SIERRA",
                "type": "text",
              },
              "tango": Object {
                "name": "TANGO",
                "type": "text",
              },
            },
          },
          "node": Object {
            "1234-1234-1234": Object {
              "color": "red",
              "name": "1234-1234-1234",
              "variables": Object {
                "1234-1234-1": Object {
                  "name": "my layout",
                  "type": "layout",
                },
                "1234-1234-2": Object {
                  "name": "my category",
                  "type": "categorical",
                },
                "1234-1234-3": Object {
                  "name": "my boolean",
                  "type": "boolean",
                },
                "1234-1234-4": Object {
                  "name": "other boolean 2",
                  "type": "boolean",
                },
                "1234-1234-5": Object {
                  "name": "other boolean 4",
                  "type": "boolean",
                },
              },
            },
            "bar": Object {
              "color": "blue",
              "name": "bar",
              "variables": Object {
                "1234-1234-4": Object {
                  "name": "1234-1234-4",
                  "type": "text",
                },
                "1234-1234-5": Object {
                  "name": "1234-1234-5",
                  "type": "text",
                },
                "alpha": Object {
                  "component": "Text",
                  "name": "ALPHA",
                  "type": "text",
                },
                "bravo": Object {
                  "name": "BRAVO",
                  "type": "text",
                },
                "charlie": Object {
                  "name": "CHARLIE",
                  "type": "location",
                },
              },
            },
            "foo": Object {
              "color": "black",
              "name": "foo",
            },
            "pop": Object {
              "color": "black",
              "name": "pop",
            },
          },
        },
        "name": "fooProtocol",
        "schemaVersion": 4,
      },
      "type": "PROTOCOL/SET",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/SAVE",
    },
  ],
  Array [
    Object {
      "dialog": Object {
        "error": [Error: Protocol.stages should NOT have fewer than 1 items 

],
        "onCancel": [Function],
        "onConfirm": [Function],
        "type": "Error",
      },
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOL/OPEN_DIALOG",
    },
  ],
  Array [
    Object {
      "destinationPath": "/dev/null/fake/user/protocol/path",
      "protocol": Object {
        "assetManifest": Object {
          "1234-asset-1": Object {
            "source": "1234-4567-asset-1",
          },
          "1234-asset-2": Object {
            "source": "1234-4567-asset-2",
          },
          "1234-asset-3": Object {
            "source": "1234-4567-asset-3",
          },
          "1234-asset-4": Object {
            "source": "1234-4567-asset-4",
          },
          "1234-asset-5": Object {
            "source": "1234-4567-asset-5",
          },
        },
        "codebook": Object {
          "edge": Object {
            "1234-5": Object {
              "color": "blue",
              "name": "an edge",
            },
          },
          "ego": Object {
            "variables": Object {
              "sierra": Object {
                "name": "SIERRA",
                "type": "text",
              },
              "tango": Object {
                "name": "TANGO",
                "type": "text",
              },
            },
          },
          "node": Object {
            "1234-1234-1234": Object {
              "color": "red",
              "name": "1234-1234-1234",
              "variables": Object {
                "1234-1234-1": Object {
                  "name": "my layout",
                  "type": "layout",
                },
                "1234-1234-2": Object {
                  "name": "my category",
                  "type": "categorical",
                },
                "1234-1234-3": Object {
                  "name": "my boolean",
                  "type": "boolean",
                },
                "1234-1234-4": Object {
                  "name": "other boolean 2",
                  "type": "boolean",
                },
                "1234-1234-5": Object {
                  "name": "other boolean 4",
                  "type": "boolean",
                },
              },
            },
            "bar": Object {
              "color": "blue",
              "name": "bar",
              "variables": Object {
                "1234-1234-4": Object {
                  "name": "1234-1234-4",
                  "type": "text",
                },
                "1234-1234-5": Object {
                  "name": "1234-1234-5",
                  "type": "text",
                },
                "alpha": Object {
                  "component": "Text",
                  "name": "ALPHA",
                  "type": "text",
                },
                "bravo": Object {
                  "name": "BRAVO",
                  "type": "text",
                },
                "charlie": Object {
                  "name": "CHARLIE",
                  "type": "location",
                },
              },
            },
            "foo": Object {
              "color": "black",
              "name": "foo",
              "variables": Object {},
            },
            "pop": Object {
              "color": "black",
              "name": "pop",
              "variables": Object {},
            },
          },
        },
        "lastModified": "2017-06-27T00:00:00.000Z",
        "name": "fooProtocol",
        "schemaVersion": 4,
        "stages": Array [],
      },
      "type": "PROTOCOLS/SAVE_SUCCESS",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/BUNDLE",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/home/documents/1234-active-protocol",
      "type": "PROTOCOLS/BUNDLE_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/SAVING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;

exports[`protocols saveAndBundleProtocol() triggers save and export actions 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/SAVING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/PREFLIGHT",
    },
  ],
  Array [
    Object {
      "meta": null,
      "protocol": Object {
        "assetManifest": Object {
          "1234-asset-1": Object {
            "source": "1234-4567-asset-1",
          },
          "1234-asset-2": Object {
            "source": "1234-4567-asset-2",
          },
          "1234-asset-3": Object {
            "source": "1234-4567-asset-3",
          },
          "1234-asset-4": Object {
            "source": "1234-4567-asset-4",
          },
          "1234-asset-5": Object {
            "source": "1234-4567-asset-5",
          },
        },
        "codebook": Object {
          "edge": Object {
            "1234-5": Object {
              "color": "blue",
              "name": "an edge",
            },
          },
          "ego": Object {
            "variables": Object {
              "sierra": Object {
                "name": "SIERRA",
                "type": "text",
              },
              "tango": Object {
                "name": "TANGO",
                "type": "text",
              },
            },
          },
          "node": Object {
            "1234-1234-1234": Object {
              "color": "red",
              "name": "1234-1234-1234",
              "variables": Object {
                "1234-1234-1": Object {
                  "name": "my layout",
                  "type": "layout",
                },
                "1234-1234-2": Object {
                  "name": "my category",
                  "type": "categorical",
                },
                "1234-1234-3": Object {
                  "name": "my boolean",
                  "type": "boolean",
                },
                "1234-1234-4": Object {
                  "name": "other boolean 2",
                  "type": "boolean",
                },
                "1234-1234-5": Object {
                  "name": "other boolean 4",
                  "type": "boolean",
                },
              },
            },
            "bar": Object {
              "color": "blue",
              "name": "bar",
              "variables": Object {
                "1234-1234-4": Object {
                  "name": "1234-1234-4",
                  "type": "text",
                },
                "1234-1234-5": Object {
                  "name": "1234-1234-5",
                  "type": "text",
                },
                "alpha": Object {
                  "component": "Text",
                  "name": "ALPHA",
                  "type": "text",
                },
                "bravo": Object {
                  "name": "BRAVO",
                  "type": "text",
                },
                "charlie": Object {
                  "name": "CHARLIE",
                  "type": "location",
                },
              },
            },
            "foo": Object {
              "color": "black",
              "name": "foo",
            },
            "pop": Object {
              "color": "black",
              "name": "pop",
            },
          },
        },
        "name": "fooProtocol",
        "schemaVersion": 4,
        "stages": Array [
          Object {
            "form": Object {
              "fields": Array [
                Object {
                  "prompt": "message",
                  "variable": "alpha",
                },
                Object {
                  "prompt": "message",
                  "variable": "1234-1234-4",
                },
              ],
            },
            "id": "bazz",
            "label": "bazz",
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "bar",
            },
            "type": "NameGenerator",
          },
          Object {
            "form": Object {
              "fields": Array [
                Object {
                  "prompt": "message",
                  "variable": "bravo",
                },
                Object {
                  "prompt": "message",
                  "variable": "1234-1234-5",
                },
              ],
            },
            "id": "pip",
            "label": "pip",
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "bar",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "buzz",
            "label": "buzz",
            "prompts": Array [
              Object {
                "id": "fizz",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "foo",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "foxtrot",
            "label": "foxtrot",
            "prompts": Array [
              Object {
                "id": "golf",
                "text": "not a prompt",
              },
              Object {
                "id": "hotel",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "pop",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "alpha",
            "label": "alpha",
            "prompts": Array [
              Object {
                "edges": Object {
                  "create": "1234-5",
                  "display": Array [
                    "1234-5",
                    "delta",
                    "echo",
                  ],
                },
                "id": "bravo",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "edge",
              "type": "1234-5",
            },
            "type": "Sociogram",
          },
          Object {
            "id": "mike",
            "label": "mike",
            "prompts": Array [
              Object {
                "id": "oscar",
                "text": "not a prompt",
              },
            ],
            "type": "Sociogram",
          },
          Object {
            "background": Object {
              "image": "1234-asset-5",
            },
            "dataSource": "1234-asset-3",
            "id": "assets",
            "items": Array [
              Object {
                "content": "1234-asset-4",
                "id": "itemId",
                "type": "asset",
              },
            ],
            "label": "assets",
            "panels": Array [
              Object {
                "dataSource": "1234-asset-2",
                "id": "panelId",
                "title": "",
              },
            ],
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "type": "Information",
          },
        ],
      },
      "type": "PROTOCOL/SET",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/SAVE",
    },
  ],
  Array [
    Object {
      "destinationPath": "/dev/null/fake/user/protocol/path",
      "protocol": Object {
        "assetManifest": Object {
          "1234-asset-1": Object {
            "source": "1234-4567-asset-1",
          },
          "1234-asset-2": Object {
            "source": "1234-4567-asset-2",
          },
          "1234-asset-3": Object {
            "source": "1234-4567-asset-3",
          },
          "1234-asset-4": Object {
            "source": "1234-4567-asset-4",
          },
          "1234-asset-5": Object {
            "source": "1234-4567-asset-5",
          },
        },
        "codebook": Object {
          "edge": Object {
            "1234-5": Object {
              "color": "blue",
              "name": "an edge",
            },
          },
          "ego": Object {
            "variables": Object {
              "sierra": Object {
                "name": "SIERRA",
                "type": "text",
              },
              "tango": Object {
                "name": "TANGO",
                "type": "text",
              },
            },
          },
          "node": Object {
            "1234-1234-1234": Object {
              "color": "red",
              "name": "1234-1234-1234",
              "variables": Object {
                "1234-1234-1": Object {
                  "name": "my layout",
                  "type": "layout",
                },
                "1234-1234-2": Object {
                  "name": "my category",
                  "type": "categorical",
                },
                "1234-1234-3": Object {
                  "name": "my boolean",
                  "type": "boolean",
                },
                "1234-1234-4": Object {
                  "name": "other boolean 2",
                  "type": "boolean",
                },
                "1234-1234-5": Object {
                  "name": "other boolean 4",
                  "type": "boolean",
                },
              },
            },
            "bar": Object {
              "color": "blue",
              "name": "bar",
              "variables": Object {
                "1234-1234-4": Object {
                  "name": "1234-1234-4",
                  "type": "text",
                },
                "1234-1234-5": Object {
                  "name": "1234-1234-5",
                  "type": "text",
                },
                "alpha": Object {
                  "component": "Text",
                  "name": "ALPHA",
                  "type": "text",
                },
                "bravo": Object {
                  "name": "BRAVO",
                  "type": "text",
                },
                "charlie": Object {
                  "name": "CHARLIE",
                  "type": "location",
                },
              },
            },
            "foo": Object {
              "color": "black",
              "name": "foo",
              "variables": Object {},
            },
            "pop": Object {
              "color": "black",
              "name": "pop",
              "variables": Object {},
            },
          },
        },
        "lastModified": "2017-06-27T00:00:00.000Z",
        "name": "fooProtocol",
        "schemaVersion": 4,
        "stages": Array [
          Object {
            "form": Object {
              "fields": Array [
                Object {
                  "prompt": "message",
                  "variable": "alpha",
                },
                Object {
                  "prompt": "message",
                  "variable": "1234-1234-4",
                },
              ],
            },
            "id": "bazz",
            "label": "bazz",
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "bar",
            },
            "type": "NameGenerator",
          },
          Object {
            "form": Object {
              "fields": Array [
                Object {
                  "prompt": "message",
                  "variable": "bravo",
                },
                Object {
                  "prompt": "message",
                  "variable": "1234-1234-5",
                },
              ],
            },
            "id": "pip",
            "label": "pip",
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "bar",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "buzz",
            "label": "buzz",
            "prompts": Array [
              Object {
                "id": "fizz",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "foo",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "foxtrot",
            "label": "foxtrot",
            "prompts": Array [
              Object {
                "id": "golf",
                "text": "not a prompt",
              },
              Object {
                "id": "hotel",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "node",
              "type": "pop",
            },
            "type": "NameGenerator",
          },
          Object {
            "id": "alpha",
            "label": "alpha",
            "prompts": Array [
              Object {
                "edges": Object {
                  "create": "1234-5",
                  "display": Array [
                    "1234-5",
                    "delta",
                    "echo",
                  ],
                },
                "id": "bravo",
                "text": "not a prompt",
              },
            ],
            "subject": Object {
              "entity": "edge",
              "type": "1234-5",
            },
            "type": "Sociogram",
          },
          Object {
            "id": "mike",
            "label": "mike",
            "prompts": Array [
              Object {
                "id": "oscar",
                "text": "not a prompt",
              },
            ],
            "type": "Sociogram",
          },
          Object {
            "background": Object {
              "image": "1234-asset-5",
            },
            "dataSource": "1234-asset-3",
            "id": "assets",
            "items": Array [
              Object {
                "content": "1234-asset-4",
                "id": "itemId",
                "type": "asset",
              },
            ],
            "label": "assets",
            "panels": Array [
              Object {
                "dataSource": "1234-asset-2",
                "id": "panelId",
                "title": "",
              },
            ],
            "prompts": Array [
              Object {
                "id": "promptId",
                "text": "not a prompt",
              },
            ],
            "type": "Information",
          },
        ],
      },
      "type": "PROTOCOLS/SAVE_SUCCESS",
    },
  ],
  Array [
    Object {
      "type": "PROTOCOLS/BUNDLE",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/home/documents/1234-active-protocol",
      "type": "PROTOCOLS/BUNDLE_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/SAVING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;

exports[`protocols unbundleAndLoadProtocol() when the protocol is invalid it dispatches an error when protocol is invalid (but still imports it) 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PREVIEW/CLOSE_PREVIEW",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/invalid",
      "type": "PROTOCOLS/UNBUNDLE",
    },
  ],
  Array [
    Object {
      "dialog": Object {
        "error": [Error: Protocol.stages should NOT have fewer than 1 items 

],
        "onCancel": [Function],
        "onConfirm": [Function],
        "type": "Error",
      },
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOL/OPEN_DIALOG",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/invalid",
      "type": "PROTOCOLS/UNBUNDLE_SUCCESS",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "advanced": true,
      "filePath": "/dev/null/mock/path/invalid",
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOLS/REGISTER",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;

exports[`protocols unbundleAndLoadProtocol() when the schema version is greater than the app version it notifies user to upgrade the app 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PREVIEW/CLOSE_PREVIEW",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/newer-protocol",
      "type": "PROTOCOLS/UNBUNDLE",
    },
  ],
  Array [
    Object {
      "dialog": Object {
        "message": <React.Fragment>
          <p>
            This protocol is not compatible with the current version of Architect.
          </p>
          <p>
            In order to open it, you will need to install a version of Architect that supports schema version 
            5
            .
          </p>
        </React.Fragment>,
        "onCancel": [Function],
        "onConfirm": [Function],
        "title": "Protocol not compatible with current version",
        "type": "UserError",
      },
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOL/OPEN_DIALOG",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;

exports[`protocols unbundleAndLoadProtocol() when the schema version is less than the app version it attempts to upgrade protocol and then imports it 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PREVIEW/CLOSE_PREVIEW",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/older-protocol",
      "type": "PROTOCOLS/UNBUNDLE",
    },
  ],
  Array [
    Object {
      "dialog": Object {
        "confirmLabel": "Create upgraded copy",
        "message": <React.Fragment>
          <p>
            This protocol uses an out-dated schema (schema version "
            3
            "), but can be upgraded to work with this version of Architect (schema version "
            4
            ").
          </p>
          <p>
            An upgraded copy of the protocol will be created and then opened.
          </p>
        </React.Fragment>,
        "onCancel": [Function],
        "onConfirm": [Function],
        "title": "Would you like to upgrade the protocol?",
        "type": "Confirm",
      },
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOL/OPEN_DIALOG",
    },
  ],
  Array [
    Object {
      "filePath": "filename.canvas",
      "type": "PROTOCOLS/UNBUNDLE_SUCCESS",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "advanced": true,
      "filePath": "filename.canvas",
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOLS/REGISTER",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;

exports[`protocols unbundleAndLoadProtocol() when the schema version is the same as the app version it imports the protocol 1`] = `
Array [
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/BUSY",
    },
  ],
  Array [
    Object {
      "type": "PREVIEW/CLOSE_PREVIEW",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/matching-protocol",
      "type": "PROTOCOLS/UNBUNDLE",
    },
  ],
  Array [
    Object {
      "filePath": "/dev/null/mock/path/matching-protocol",
      "type": "PROTOCOLS/UNBUNDLE_SUCCESS",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "advanced": true,
      "filePath": "/dev/null/mock/path/matching-protocol",
      "id": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
      "type": "PROTOCOLS/REGISTER",
      "workingPath": "/dev/null/get/electron/path/protocols/809895df-bbd7-4c76-ac58-e6ada2625f9b",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS/LOADING",
      "type": "STATUS/READY",
    },
  ],
  Array [
    Object {
      "payload": "PROTOCOLS",
      "type": "STATUS/READY",
    },
  ],
]
`;
